### PR TITLE
Tx metadata API integration tests

### PR DESCRIPTION
**Pull request by rvl**
### Issue Number

ADP-307 / #2074 / #2073.

### Overview

- API test case for posting a transaction with metadata.
- Test creating a transaction with invalid metadata.
- Test creating a transaction that's too large due to metadata.
- Test that the transaction fee estimate is higher when there's metadata in the transaction.

### Comments

- Based on PR #2104 branch.

### CODE REVIEW

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 1]
```diff
@@ -229,6 +241,7 @@ spec = do
                 between (feeMin + amt, feeMax + amt)
             , expectField (#direction . #getApiT) (`shouldBe` Outgoing)
             , expectField (#status . #getApiT) (`shouldBe` Pending)
+            , expectField (#metadata . #getApiTxMetadata) (`shouldBe` Nothing)
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:247]: :+1:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 2]
```diff
@@ -533,6 +546,102 @@ spec = do
...
+        eventually "metadata is confirmed in transaction list" $ do
+            let link = Link.listTransactions @'Shelley wa
+            rb <- request @([ApiTransaction n]) ctx link Default Empty
+            verify rb
+                [ expectResponseCode HTTP.status200
+                , expectListField 0 (#status . #getApiT) (`shouldBe` InLedger)
+                , expectListField 0
+                    (#metadata . #getApiTxMetadata)
+                    (`shouldBe` Just (ApiT expected))
+                ]
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:585]: :+1:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 3]
```diff
@@ -533,6 +546,102 @@ spec = do
...
+
+    it "TRANS_CREATE_11 - Transaction with invalid metadata" $ \ctx -> do
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = (1 :: Natural)
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = Aeson.object ["1" .= T.replicate 65 "a"]
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:593]: :+1:

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 4]
```diff
@@ -533,6 +546,102 @@ spec = do
...
+        (wa, wb) <- (,) <$> fixtureWallet ctx <*> fixtureWallet ctx
+        let amt = (1 :: Natural)
+
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        -- This will encode to at least 64k of CBOR.
+        let txMeta = Aeson.object
+                [ (toText @Int i, Aeson.String (T.replicate 64 "a"))
+                | i <- [0..1023] ]
+        let payload = addTxMetadata txMeta basePayload
```
Comment (KtorZ) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs:613]: :+1:

**Review by KtorZ — APPROVED** [review]
LGTM

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 5]
```diff
@@ -533,6 +546,102 @@ spec = do
...
+        basePayload <- mkTxPayload ctx wb amt fixturePassphrase
+
+        let txMeta = Aeson.object ["1" .= T.replicate 65 "a"]
+        let payload = addTxMetadata txMeta basePayload
+
+        r <- request @(ApiTransaction n) ctx
+            (Link.createTransaction @'Shelley wa) Default payload
+
+        expectResponseCode @IO HTTP.status400 r
+        expectErrorMessage "fixme: api validation error message" r
```
Comment (piotr-iohk) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: This message is unhappy, it's crying "fixme" ;)

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs [hunk 6]
```diff
@@ -533,6 +546,102 @@ spec = do
...
+        let txMeta = Aeson.object
+                [ (toText @Int i, Aeson.String (T.replicate 64 "a"))
+                | i <- [0..1023] ]
+        let payload = addTxMetadata txMeta basePayload
+
+        r <- request @(ApiTransaction n) ctx
+            (Link.createTransaction @'Shelley wa) Default payload
+
+        expectResponseCode @IO HTTP.status400 r
+        expectErrorMessage "fixme: transaction rejected message" r
```
Comment (piotr-iohk) [inline lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: so does this one

**Review by piotr-iohk — APPROVED** [review]
lgtm.

There are few other cases come to my mind, pertaining to invalid metadata but maybe a bit more subtle:
 - top-level index > 2^64-1
```
{ 999999999999999999999999999999999999999999999999999999999999999999999999999999999: "ziemniak" }
```
 - hex that is not hex
```
{ 1: { "hex": "ąćó" } }
```
- map that is not map
```
{ "1": [ [ "k1", "v1", "v2" ] ] }
```

I wonder would that produce invalid metadata error?

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 5 above)
Comment (rvl) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Yes, I needed to run the tests first to find out what error message to expect. :slightly_smiling_face: 
Fixed now.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 6 above)
Comment (rvl) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: See latest update of the branch.
Turns out the error message for oversized transactions is not very good.
I think we should fix the error message - but in another PR.

Comment (KtorZ) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Too late. I've done it in this PR :grimacing:

**Comment by KtorZ** [comment]
bors retry

**Comment by KtorZ** [comment]
Buildkite timedout but Hydra passed. Merging this, but we likely need to increase buildkite timeout (or remove integration tests from buildkite since they're already part of hydra anyway).

**Comment by rvl** [comment]
@piotr-iohk Thanks for the suggestions about test coverage for invalid metadata cases. I added a note to the QA section of #2073 about this.

These cases are covered in `cardano-node/cardano-api/test/Test/Cardano/Api/MetaData.hs` by property tests. So we only need one invalid metadata test (I chose one with a string that was too long) to test the API error path.

File: lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs (hunk 6 above)
Comment (rvl) [reply lib/core-integration/src/Test/Integration/Scenario/API/Shelley/Transactions.hs]: Thanks

**Pull request merged**